aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+          ));
+  } else { // inference
+    reserve = input.type().toScalarType(kByte).tensor();
+    CUDNN_CHECK(cudnnRNNForwardInference(
+          handle,
+          descs.rnn_desc.desc(),
+          fn.tensors.seq_length,
+          x_descs_arr.data(), x.data_ptr(),
+          descs.hx_desc.desc(), hx.data_ptr(),
+          descs.cx_desc.desc(), cx.defined() ? cx.data_ptr() : nullptr,
+          w_desc.desc(), weight_buf.data_ptr(),
+          y_descs_arr.data(), y.data_ptr(),
+          descs.hy_desc.desc(), hy.data_ptr(),
+          descs.cy_desc.desc(), cy.defined() ? cy.data_ptr() : nullptr,
+          workspace.data_ptr(), workspace.size(0)
+          ));
+
+  }
+
+  if (batch_first && !is_input_packed) {
+    output.transpose_(0, 1);
+  }
+
+  return std::make_tuple(output, hy, cy, reserve, weight_buf);
+}
+
+std::tuple<Tensor, Tensor, Tensor> _cudnn_rnn_backward_grad(
+    const Tensor& input_r, const Tensor& weight_buf, const Tensor& hx, const Tensor& cx,
+    const Tensor& output_r, const Tensor& grad_output_r, const Tensor& grad_hy,
+    const Tensor& grad_cy,
+    int64_t fn_mode, int64_t fn_hidden_size,
+    int64_t fn_num_layers, bool batch_first, double fn_dropout,
+    bool fn_train, bool fn_bidirectional, IntList fn_batch_sizes,
+    const Tensor& fn_dropout_state, const Tensor& fn_reserve
+    ) {
+
+  auto input = input_r;
+  auto grad_output = grad_output_r;
+  auto output = output_r;
+
+  RNNParams fn;
+  fn.rnn.set_mode(fn_mode);
+  fn.rnn.hidden_size = fn_hidden_size;
+  fn.rnn.num_layers = fn_num_layers;
+  fn.rnn.set_bidirectional(fn_bidirectional);
+  fn.rnn.datatype = getCudnnDataType(input);
+  fn.dropout.set(fn_train, fn_dropout, fn_dropout_state);
+  fn.tensors.set(input.sizes(), fn_batch_sizes, batch_first);
+
+  // TODO: Set device to input
+  auto handle = getCudnnHandle();
+
+  if (fn.rnn.mode != CUDNN_LSTM) {
+    if (cx.defined()) {
+      throw std::runtime_error("rnn: illegal defined cx for non-LSTM RNN");
+    }
+  }
+
+  auto is_input_packed = fn_batch_sizes.size() != 0;
+  if (batch_first && !is_input_packed) {
+    input = input.transpose(0, 1);
+    grad_output = grad_output.transpose(0, 1);
+    output = output.transpose(0, 1);
+  }
+
+  auto input_size = _input_size(fn.tensors);
+  auto hidden_size = _hidden_size(fn.rnn, fn.tensors);
+  auto output_size = _output_size(fn.rnn, fn.tensors);
+
+  AT_ASSERT(hx.is_contiguous(), "hx.is_contiguous()");
+  AT_ASSERT(!cx.defined() || cx.is_contiguous(), "!cx or cx.is_contiguous()");
+
+  auto x = input.contiguous();
+  auto dy = grad_output.contiguous();
+  auto y = output;
+  auto w = weight_buf;
+  auto dx = input.type().tensor(input.sizes()); // TODO: more compact way of saying this
+  auto dhy = grad_hy.contiguous().view(hidden_size);
+  auto dcy = grad_cy.defined() ? grad_cy.contiguous().view(hidden_size) : Tensor();
+  auto dhx = hx.type().tensor(hidden_size);
+  auto dcx = cx.defined() ? cx.type().tensor(hidden_size) : hx.type().tensor(); // Boooh
```

Comment:
?

Reply:
When I originally wrote this I thought that maybe it would be illegal to return an undefined tensor here, but actually it should always be the case that `dcx` is never requested when `cx` is undefined, so we are allowed to return undefined. I've added an assert to this effect.